WcaOnRails/spec/models/competition_event_spec.rb

```diff
@@ -0,0 +1,60 @@
+# frozen_string_literal: true
+require 'rails_helper'
+
+RSpec.describe CompetitionEvent do
+  let(:competition) { FactoryGirl.create :competition, event_ids: %w(333 444) }
+  let(:competition_event) { competition.competition_events.find_by_event_id("333") }
+
+  it "it's okay if there are no rounds yet" do
+    expect(competition_event).to be_valid
+  end
+
+  context "nth_round" do
+    it "must start at 1" do
+      competition_event.rounds_attributes = [
+        { nth_round: 3, format_id: "a" },
+      ]
+      expect(competition_event).to be_invalid
+      expect(competition_event.errors[:rounds]).to eq ["[3] is wrong"]
+    end
+
+    it "must be contiguous" do
+      competition_event.rounds_attributes = [
+        { nth_round: 1, format_id: "a" },
+        { nth_round: 3, format_id: "a" },
+      ]
+      expect(competition_event).to be_invalid
+      expect(competition_event.errors[:rounds]).to eq ["[1, 3] is wrong"]
+    end
+
+    it "cannot have duplicates" do
+      competition_event.rounds_attributes = [
+        { nth_round: 1, format_id: "a" },
+        { nth_round: 2, format_id: "a" },
+        { nth_round: 2, format_id: "a" },
+        { nth_round: 3, format_id: "a" },
+      ]
+      expect(competition_event).to be_invalid
+      expect(competition_event.errors[:rounds]).to eq ["[1, 2, 2, 3] is wrong"]
+    end
+
+    it "cannot remove a round in the middle" do
+      competition_event.rounds_attributes = [
+        { nth_round: 1, format_id: "a" },
+        { nth_round: 2, format_id: "a" },
+        { nth_round: 3, format_id: "a" },
+      ]
+      competition_event.save!
+      expect(competition_event).to be_valid
+      expect(competition_event.rounds.length).to eq 3
+
+      competition_event.rounds_attributes = [
+        { id: competition_event.rounds[0].id },
+        { id: competition_event.rounds[1].id, _destroy: "1" },
+        { id: competition_event.rounds[2].id },
+      ]
+      expect(competition_event).to be_invalid
+      expect(competition_event.errors[:rounds]).to eq ["[1, 3] is wrong"]
+    end
+  end
+end
```

Comment:
Maybe it's just me, but I kinda don't like the redundancy across those specs. I'd suggest something like this
*Note1: I didn't check if this runs, so sorry if there's any typo.*
*Note2: If this looks like an overkill then feel free to ignore me!*

```ruby
# frozen_string_literal: true
require 'rails_helper'

RSpec.describe CompetitionEvent do
  let(:competition) { FactoryGirl.create :competition, event_ids: %w(333 444) }
  let(:competition_event) { competition.competition_events.find_by_event_id("333") }

  it "it's okay if there are no rounds yet" do
    expect(competition_event).to be_valid
  end

  context "nth_round" do
    it "must start at 1" do
      expect_rounds of_order: [3], to_give_errors: ["[3] is wrong"]
    end

    it "must be contiguous" do
      expect_rounds of_order: [1, 3], to_give_errors: ["[1, 3] is wrong"]
    end

    it "cannot have duplicates" do
      expect_rounds of_order: [1, 2, 2, 3], to_give_errors: ["[1, 2, 2, 3] is wrong"]
    end

    it "cannot remove a round in the middle" do
      expect_rounds of_order: [1, 2, 3], to_give_errors: nil
      competition_event.save!
      expect(competition_event.rounds.length).to eq 3

      competition_event.rounds_attributes = competition_event.rounds.map { |r| { id: r.id } }
      competition_event.rounds_attributes[1]._destroy = "1"

      expect(competition_event).to be_invalid
      expect(competition_event.errors[:rounds]).to eq ["[1, 3] is wrong"]
    end
  end

  def expect_rounds(of_order:, to_give_errors:)
    competition_event.rounds_attributes = of_order.map { |n| { nth_round: n, format_id: "a" } }
    expect(competition_event).to(to_give_errors ? be_invalid : be_valid)
    expect(competition_event.errors[:rounds]).to eq to_give_errors
  end
end
```